### PR TITLE
develop

### PR DESCRIPTION
**Pull request by h8718**
## Description

WiP

### CODE REVIEW

**Review by alankm — COMMENTED** [review]
Why did you add a new Dockerfile to the root of the project? And why did you change the make targets to point at this new file instead of the file they were pointing at before?

**Comment by h8718** [comment]
> Why did you add a new Dockerfile to the root of the project? And why did you change the make targets to point at this new file instead of the file they were pointing at before?

### Dockerfile Modifications and CI/CD Process Refinements:

    **Why**? To ensure that our CI/CD processes align with our infrastructure choices, a new Dockerfile was introduced to root, motivated by the need to leverage ghcr as our image registry, given the complexity with GitHub actions and Docker commands. Besides, the base of the Dockerfile has changed from Ubuntu to Alpine, enhancing performance.
    **Benefits**: Performance enhancement and alignment with GitHub's CI/CD system.
    **Considerations**: We need to be aware that currently the pipeline does not verify before pushing the image passes all pipeline checks.

**Comment by h8718** [comment]
**Included PRs in this Branch**:

- [ ] [PR#1020 Adds the event delay feature](https://github.com/direktiv/direktiv/pull/1020)
- [ ] [PR#1024 Adds a missed property for eventstreams](https://github.com/direktiv/direktiv/pull/1024)
- [ ] [PR#1023 Fixes a bug regarding workflow-names](https://github.com/direktiv/direktiv/pull/1023)
- [ ] [PR#1010 Code cleanup and minor improvements of the new Bus-System](https://github.com/direktiv/direktiv/pull/1010)
- [ ] [PR#997 Change base-image](https://github.com/direktiv/direktiv/pull/997)

**Other key Adjustments**:

- **CI/CD Pipeline Modifications**: We've refined our CI/CD pipeline, laying the groundwork for automated deployments in our development environment.

**Rationale for Consolidation**:

The decision to consolidate these PRs arose from the challenges posed by managing multiple interconnected changes. By combining them, we aim to:

1. Reduce the overhead and complexities of synchronizing individual changes.
2. Provide a more coherent and comprehensive review experience for collaborators and contributors.
3. Ensure consistent and synergistic modifications when they are integrated into the main branch.
4. Facilitate faster deployment of bugfixes into the development environment, to prevent blocking other developers

File: pkg/cluster/node-finder.go [hunk 1]
```diff
@@ -1,17 +1,19 @@
...
+	"time"
 )
 
 // Nodefinders have to return all ips/addr of the serf nodes available on startup
 // the minimum is to return one ip/addr for serf to form a cluster.
 type NodeFinder interface {
-	GetNodes() ([]string, error)
-	GetAddr() (string, error)
+	GetNodes(ctx context.Context) ([]string, error)
+	GetAddr(ctx context.Context, nodeID string) (string, error)
```
Comment (alankm) [inline pkg/cluster/node-finder.go:16]: By adding the nodeID argument I think the purpose of GetAddr has changed pretty significantly. The original purpose was to get the addr of the self. Why was this changed?

Comment (alankm) [reply pkg/cluster/node-finder.go:16]: It looks like both implementations of GetAddr discard this argument as well, so that seems extra strange.

File: pkg/cluster/node.go [hunk 2]
```diff
@@ -4,71 +4,78 @@
...
-	if err != nil {
-		return nil, fmt.Errorf("failed to create nsq bus: %w", err)
-	}
-	go func() {
-		e := node.bus.start()
-		if e != nil {
-			panic("can not start nsq bus")
-		}
-	}()
+	ctx, cancel := context.WithCancel(ctx)
```
Comment (alankm) [inline pkg/cluster/node.go:63]: I believe cancelling contexts is always necessary. This function doesn't cancel if everything returns successfully.

File: pkg/cluster/node_test.go [hunk 3]
```diff
@@ -44,25 +66,28 @@ func TestNewNode(t *testing.T) {
 }
 
 func rightNumber(nodes []*Node) bool {
-	for i := 0; i < len(nodes); i++ {
-		mnodes := nodes[i].serfServer.Members()
-		count := len(mnodes)
-		for j := 0; j < len(mnodes); j++ {
-			if mnodes[j].Status == serf.StatusLeft {
-				count--
+	for _, node := range nodes {
```
Comment (alankm) [inline pkg/cluster/node_test.go:69]: I spent a lot of time designing this function. I'm curious to hear why you changed it.

File: pkg/flow/bytedata/data.go [hunk 4]
```diff
@@ -223,6 +222,7 @@ func ConvertLogMsgForOutput(a []*logengine.LogEntry) ([]*grpc.Log, error) {
 	return results, nil
 }
 
+// TODO test this!!!
```
Comment (alankm) [inline pkg/flow/bytedata/data.go:225]: TODO
